src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Utilities/NullableHelpers/NullableHelpers.cs

```diff
@@ -0,0 +1,87 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Operations;
+using Roslyn.Utilities;
+
+#nullable enable
+
+namespace Microsoft.CodeAnalysis
+{
+    internal static class NullableHelpers
+    {
+        /// <summary>
+        /// Given an encapsulating operation, returns true if the symbol passed in
+        /// is ever assigned a possibly null value as determined by nullable flow state. Returns
+        /// null if no references are found, letting the caller determine what to do with that information
+        /// </summary>
+        public static bool? IsSymbolAssignedPossiblyNullValue(SemanticModel semanticModel, IOperation? containingOperation, ISymbol? symbol)
+        {
+            if (containingOperation is null || symbol is null)
+            {
+                return null;
+            }
+
+            var references = containingOperation.DescendantsAndSelf()
+                .Where(o => IsSymbolReferencedByOperation(o, symbol, allowNullInitializer: false));
+
+            var hasReference = false;
+
+            foreach (var reference in references)
+            {
+                hasReference = true;
+
+                // foreach statements are handled special because the iterator is not assignable, so the elementtype 
+                // annotation is accurate for determining if the loop declaration has a reference that allows the symbol
+                // to be null
+                if (reference is IForEachLoopOperation forEachLoop)
+                {
+                    var foreachInfo = semanticModel.GetForEachStatementInfo((CommonForEachStatementSyntax)forEachLoop.Syntax);
+
+                    // Use NotAnnotated here to keep both Annotated and None (oblivious) treated the same, since
+                    // this is directly looking at the annotation and not the flow state
+                    if (foreachInfo.ElementType.NullableAnnotation != NullableAnnotation.NotAnnotated)
+                    {
+                        return true;
+                    }
+
+                    continue;
+                }
+
+                var syntax = reference switch
+                {
+                    IVariableDeclaratorOperation variableDeclarator => variableDeclarator.GetVariableInitializer().Value.Syntax,
+                    _ => reference.Syntax
+                };
```

Comment:
```suggestion
                var syntax = reference is IVariableDeclaratorOperation variableDeclarator
                    ? variableDeclarator.GetVariableInitializer().Value.Syntax
                    : reference.Syntax
```

(a switch with two branches seems unnecessary).